opentreemap/treemap/instance.py

```diff
@@ -768,9 +771,39 @@ def raise_errors(errors):
         if errors:
             raise_errors(errors)
 
-    def save(self, *args, **kwargs):
+    @transaction.atomic
+    def _bulk_assign_task_reference_numbers(self, tasks):
+        instance = Instance.objects.select_for_update().get(id=self.id)
+        for task in tasks:
+            task.reference_number = instance.task_sequence_number
+            instance.task_sequence_number += 1
+        instance.save(update_fields=['task_sequence_number'])
+
+    @transaction.atomic
+    def _bulk_assign_work_order_reference_numbers(self, work_orders):
+        instance = Instance.objects.select_for_update().get(id=self.id)
+        for work_order in work_orders:
+            work_order.reference_number = instance.work_order_sequence_number
+            instance.work_order_sequence_number += 1
+        instance.save(update_fields=['work_order_sequence_number'])
+
+    def save(self, tasks=None, work_orders=None, *args, **kwargs):
+        """
+        Keyword Arguments:
+
+        tasks - List of Task models
+        work_orders - List of WorkOrder models
+        """
         self.full_clean()
 
         self.url_name = self.url_name.lower()
 
+        if tasks:
```

Comment:
This logic could very easily be moved to a utility function or classmethod. I thought putting it here would prevent a data consistency error, but I just realized that's not true. For example:

```py
# User 1
instance1 = Instance.objects.get(id=1)
# User 2
instance2 = Instance.objects.get(id=1)

# User 1
t = make_task(instance1, ...)
self.assertEqual(1, t.reference_number)
self.assertEqual(2, instance1.task_sequence_number)

# User 2
instance2.save()

# AssertionError: 2 != 1
instance3 = Instance.objects.get(id=1)
self.assertEqual(2, instance3.task_sequence_number)
```

As long as these sequence fields exist on the `Instance` model there is the potential for data consistency errors. For example, using a helper method to allocate sequence numbers inside a transaction:

```py
# User 1
instance = get_instance() # task_sequence_number=1

# User 2
instance = get_instance() # task_sequence_number=1
n = task_sequence_increment(instance, 20) # task_sequence_number=21
tasks[0].reference_number = n
tasks[1].reference_number = n + 1
# etc.

# User 1
instance.save() # task_sequence_number=1; Whoops!
```

Do you think this is a real issue we should be concerned about?

If we could define "unmanaged" fields in Django, that could help avoid this problem. Another alternative could be to use Redis?